File: Assets/Scripts/Module/Invoke/FActions/GraphicToAction.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;
namespace FAction {
    public class GraphicToAction : FAction
    {
        public enum GraphicControlType
        {
            /// <summary>
            /// 仅alpha
            /// </summary>
            OnlyAlpha = 0,
            /// <summary>
            /// 除了alpha
            /// </summary>
            ExcludeAlpha = 1,
            /// <summary>
            /// rgba
            /// </summary>
            All = 2,
        }

        public override ActionType ActionType => ActionType.GraphicTo;
        public GraphicControlType ControlType { get; private set; }
        Graphic graphic;

        Color target;
        Color start;
        float time = 0;

        float currentTime = 0;

        public void InitGraphicInfo( Graphic graphic, GraphicControlType type, Color target, float duration)
        {
            this.graphic = graphic;
            ControlType = type;

            this.target = target;
            this.time = duration;
        }

        public override void Replay()
        {
            base.Replay();
            currentTime = 0;
        }
        public override void OnStartTick()
        {
            start = graphic.color;
        }

        public override void Tick()
        {
            base.Tick();
            if (time <= 0)
            {
                SetGraphic(target);
                this.IsFinish = true;
                return;
            }


            currentTime += Time.deltaTime;
            float percent = currentTime / time;
            if (percent >= 1)
            {
                SetGraphic(target);
                this.IsFinish = true;
                return;
            }
            var lerp = Color.Lerp(start, target, percent);
            SetGraphic(lerp);
    
        }

        public override void Clear()
        {
            base.Clear();
            graphic = null;
            currentTime = 0;
            time = 0;
        }

        void SetGraphic(Color target)
        {
            var current = graphic.color;
            switch (ControlType)
            {
                case GraphicControlType.OnlyAlpha:
                    graphic.color = new Color(current.r, current.g, current.b, target.a);
                    break;
                case GraphicControlType.ExcludeAlpha:
                    graphic.color = new Color(target.r, target.g, target.b, current.a);
                    break;
                case GraphicControlType.All:
                    graphic.color = target;
                    break;
                default:
                    break;
            }
        }
    }
}

File: Assets/.lpproj/Apis/Unity/CS/FAction/GraphicToAction.lua
---@class CS.FAction.GraphicToAction : CS.FAction.FAction
CS.FAction.GraphicToAction = {}

---@property readonly CS.FAction.GraphicToAction.ActionType : CS.FAction.ActionType
CS.FAction.GraphicToAction.ActionType = nil

---@property readwrite CS.FAction.GraphicToAction.ControlType : CS.FAction.GraphicControlType
CS.FAction.GraphicToAction.ControlType = nil

---@return CS.FAction.GraphicToAction
function CS.FAction.GraphicToAction()
end

---@param graphic : CS.UnityEngine.UI.Graphic
---@param type : CS.FAction.GraphicControlType
---@param target : CS.UnityEngine.Color
---@param duration : CS.System.Single
function CS.FAction.GraphicToAction:InitGraphicInfo(graphic, type, target, duration)
end

function CS.FAction.GraphicToAction:Replay()
end

function CS.FAction.GraphicToAction:OnStartTick()
end

function CS.FAction.GraphicToAction:Tick()
end

function CS.FAction.GraphicToAction:Clear()
end